File: scripts/lint-bank.py
#!/usr/bin/env python3
"""
Bank linter for avalanche survey item banks
Validates JSON schema and business rules
"""

import json
import sys
import os
from pathlib import Path

def load_json(file_path):
    """Load and parse JSON file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        raise Exception(f"Failed to load {file_path}: {e}")

def validate_schema_basic(data, required_fields, field_name=""):
    """Basic schema validation"""
    errors = []
    
    if not isinstance(data, dict):
        errors.append(f"{field_name}: Expected object, got {type(data).__name__}")
        return errors
    
    # Check required fields
    for field in required_fields:
        if field not in data:
            errors.append(f"{field_name}.{field}: Required field missing")
    
    return errors

def validate_business_rules(bank):
    """Validate business rules for item bank"""
    errors = []
    
    # Check unique IDs across all questions
    all_ids = set()
    duplicate_ids = set()
    
    # Collect IDs from testlets
    for testlet in bank.get('testlets', []):
        for item in testlet.get('items', []):
            item_id = item.get('id')
            if item_id:
                if item_id in all_ids:
                    duplicate_ids.add(item_id)
                all_ids.add(item_id)
    
    # Collect IDs from diagnostics
    for diagnostic in bank.get('diagnostics', []):
        diagnostic_id = diagnostic.get('id')
        if diagnostic_id:
            if diagnostic_id in all_ids:
                duplicate_ids.add(diagnostic_id)
            all_ids.add(diagnostic_id)
    
    if duplicate_ids:
        errors.append(f"Duplicate question IDs found: {', '.join(duplicate_ids)}")
    
    # Check that each testlet has all four constructs
    required_constructs = {'development', 'behaviour', 'assessment', 'mitigation'}
    
    for testlet in bank.get('testlets', []):
        constructs = set()
        for item in testlet.get('items', []):
            construct = item.get('construct')
            if construct:
                constructs.add(construct)
        
        missing_constructs = required_constructs - constructs
        if missing_constructs:
            testlet_type = testlet.get('ap_type', 'unknown')
            errors.append(f"Testlet \"{testlet_type}\" missing constructs: {', '.join(missing_constructs)}")
    
    # Check that each item has a valid key present in choices
    all_questions = []
    for testlet in bank.get('testlets', []):
        all_questions.extend(testlet.get('items', []))
    all_questions.extend(bank.get('diagnostics', []))
    
    for question in all_questions:
        question_id = question.get('id', 'unknown')
        choices = question.get('choices', [])
        key = question.get('key')
        
        if choices and key:
            choice_values = [choice.get('value') for choice in choices if isinstance(choice, dict)]
            if key not in choice_values:
                errors.append(f"Question \"{question_id}\" has key \"{key}\" not found in choices: {', '.join(choice_values)}")
        
        # Check non-empty explanations
        explain = question.get('explain', '').strip()
        if not explain:
            errors.append(f"Question \"{question_id}\" has empty or missing explanation")
    
    return errors

def lint_bank(bank_path):
    """Main linting function"""
    print(f"🔍 Linting bank: {bank_path}")
    
    try:
        # Load bank
        bank = load_json(bank_path)
        
        # Basic schema validation
        schema_errors = []
        schema_errors.extend(validate_schema_basic(bank, ['schema_version', 'testlets', 'diagnostics']))
        
        # Validate testlets
        for i, testlet in enumerate(bank.get('testlets', [])):
            testlet_errors = validate_schema_basic(testlet, ['ap_type', 'label', 'items'], f"testlets[{i}]")
            schema_errors.extend(testlet_errors)
            
            # Validate items in testlet
            for j, item in enumerate(testlet.get('items', [])):
                item_errors = validate_schema_basic(item, ['id', 'construct', 'stem', 'choices', 'key', 'explain'], f"testlets[{i}].items[{j}]")
                schema_errors.extend(item_errors)
        
        # Validate diagnostics
        for i, diagnostic in enumerate(bank.get('diagnostics', [])):
            diagnostic_errors = validate_schema_basic(diagnostic, ['id', 'construct', 'stem', 'choices', 'key', 'explain'], f"diagnostics[{i}]")
            schema_errors.extend(diagnostic_errors)
        
        if schema_errors:
            print('❌ Schema validation failed:')
            for error in schema_errors:
                print(f"  - {error}")
            return False
        
        print('✅ Schema validation passed')
        
        # Validate business rules
        business_errors = validate_business_rules(bank)
        if business_errors:
            print('❌ Business rule validation failed:')
            for error in business_errors:
                print(f"  - {error}")
            return False
        
        print('✅ Business rule validation passed')
        
        # Summary
        testlet_count = len(bank.get('testlets', []))
        question_count = sum(len(t.get('items', [])) for t in bank.get('testlets', []))
        diagnostic_count = len(bank.get('diagnostics', []))
        
        print(f'📊 Summary:')
        print(f'  - Testlets: {testlet_count}')
        print(f'  - Questions: {question_count}')
        print(f'  - Diagnostics: {diagnostic_count}')
        print('🎉 All validations passed!')
        
        return True
    
    except Exception as e:
        print(f"❌ Error: {e}")
        return False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print('Usage: python3 lint-bank.py <path-to-bank.json>')
        sys.exit(1)
    
    bank_path = sys.argv[1]
    success = lint_bank(bank_path)
    sys.exit(0 if success else 1)

File: scripts/lint-bank.mjs
#!/usr/bin/env node

/**
 * Bank linter for avalanche survey item banks
 * Validates JSON schema and business rules
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Simple JSON Schema validator (basic implementation)
class SchemaValidator {
    validate(data, schema) {
        const errors = [];
        this._validateObject(data, schema, '', errors);
        return { valid: errors.length === 0, errors };
    }

    _validateObject(data, schema, path, errors) {
        if (schema.type === 'object') {
            if (typeof data !== 'object' || data === null || Array.isArray(data)) {
                errors.push(`${path}: Expected object, got ${typeof data}`);
                return;
            }

            // Check required properties
            if (schema.required) {
                for (const prop of schema.required) {
                    if (!(prop in data)) {
                        errors.push(`${path}.${prop}: Required property missing`);
                    }
                }
            }

            // Validate properties
            if (schema.properties) {
                for (const [prop, propSchema] of Object.entries(schema.properties)) {
                    if (prop in data) {
                        this._validateObject(data[prop], propSchema, path ? `${path}.${prop}` : prop, errors);
                    }
                }
            }
        } else if (schema.type === 'array') {
            if (!Array.isArray(data)) {
                errors.push(`${path}: Expected array, got ${typeof data}`);
                return;
            }

            if (schema.minItems && data.length < schema.minItems) {
                errors.push(`${path}: Array must have at least ${schema.minItems} items, got ${data.length}`);
            }

            if (schema.items) {
                data.forEach((item, index) => {
                    this._validateObject(item, schema.items, `${path}[${index}]`, errors);
                });
            }
        } else if (schema.type === 'string') {
            if (typeof data !== 'string') {
                errors.push(`${path}: Expected string, got ${typeof data}`);
                return;
            }

            if (schema.minLength && data.length < schema.minLength) {
                errors.push(`${path}: String must be at least ${schema.minLength} characters, got ${data.length}`);
            }

            if (schema.enum && !schema.enum.includes(data)) {
                errors.push(`${path}: Value "${data}" not in allowed values: ${schema.enum.join(', ')}`);
            }

            if (schema.pattern && !new RegExp(schema.pattern).test(data)) {
                errors.push(`${path}: String "${data}" does not match pattern ${schema.pattern}`);
            }
        } else if (schema.type === 'boolean') {
            if (typeof data !== 'boolean') {
                errors.push(`${path}: Expected boolean, got ${typeof data}`);
            }
        } else if (schema.type === 'integer') {
            if (!Number.isInteger(data)) {
                errors.push(`${path}: Expected integer, got ${typeof data}`);
                return;
            }

            if (schema.minimum !== undefined && data < schema.minimum) {
                errors.push(`${path}: Value ${data} is less than minimum ${schema.minimum}`);
            }
        }

        // Handle $ref (simplified)
        if (schema.$ref && schema.$ref.startsWith('#/definitions/')) {
            const defName = schema.$ref.replace('#/definitions/', '');
            const rootSchema = this._findRootSchema(schema);
            if (rootSchema?.definitions?.[defName]) {
                this._validateObject(data, rootSchema.definitions[defName], path, errors);
            }
        }
    }

    _findRootSchema(schema) {
        // Simple approach: assume the schema object has definitions at root
        let current = schema;
        while (current && !current.definitions) {
            current = current.parent || this._rootSchema;
        }
        return current;
    }

    setRootSchema(schema) {
        this._rootSchema = schema;
    }
}

/**
 * Load and parse JSON file
 */
function loadJSON(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        return JSON.parse(content);
    } catch (error) {
        throw new Error(`Failed to load ${filePath}: ${error.message}`);
    }
}

/**
 * Validate business rules for item bank
 */
function validateBusinessRules(bank) {
    const errors = [];

    // Check unique IDs across all questions
    const allIds = new Set();
    const duplicateIds = new Set();

    // Collect IDs from testlets
    for (const testlet of bank.testlets || []) {
        for (const item of testlet.items || []) {
            if (allIds.has(item.id)) {
                duplicateIds.add(item.id);
            }
            allIds.add(item.id);
        }
    }

    // Collect IDs from diagnostics
    for (const diagnostic of bank.diagnostics || []) {
        if (allIds.has(diagnostic.id)) {
            duplicateIds.add(diagnostic.id);
        }
        allIds.add(diagnostic.id);
    }

    if (duplicateIds.size > 0) {
        errors.push(`Duplicate question IDs found: ${Array.from(duplicateIds).join(', ')}`);
    }

    // Check that each testlet has all four constructs
    for (const testlet of bank.testlets || []) {
        const constructs = new Set();
        for (const item of testlet.items || []) {
            constructs.add(item.construct);
        }

        const requiredConstructs = ['development', 'behaviour', 'assessment', 'mitigation'];
        const missingConstructs = requiredConstructs.filter(c => !constructs.has(c));
        
        if (missingConstructs.length > 0) {
            errors.push(`Testlet "${testlet.ap_type}" missing constructs: ${missingConstructs.join(', ')}`);
        }
    }

    // Check that each item has a valid key present in choices
    const allQuestions = [
        ...(bank.testlets || []).flatMap(t => t.items || []),
        ...(bank.diagnostics || [])
    ];

    for (const question of allQuestions) {
        if (!question.choices || !question.key) continue;
        
        const choiceValues = question.choices.map(c => c.value);
        if (!choiceValues.includes(question.key)) {
            errors.push(`Question "${question.id}" has key "${question.key}" not found in choices: ${choiceValues.join(', ')}`);
        }
    }

    // Check non-empty explanations
    for (const question of allQuestions) {
        if (!question.explain || question.explain.trim().length === 0) {
            errors.push(`Question "${question.id}" has empty or missing explanation`);
        }
    }

    return errors;
}

/**
 * Main linting function
 */
async function lintBank(bankPath) {
    console.log(`🔍 Linting bank: ${bankPath}`);

    try {
        // Load schema
        const schemaPath = path.join(__dirname, '../packages/shared/schemas/bank.schema.json');
        const schema = loadJSON(schemaPath);
        
        // Load bank
        const bank = loadJSON(bankPath);

        // Validate schema
        const validator = new SchemaValidator();
        validator.setRootSchema(schema);
        const schemaResult = validator.validate(bank, schema);

        if (!schemaResult.valid) {
            console.error('❌ Schema validation failed:');
            schemaResult.errors.forEach(error => console.error(`  - ${error}`));
            return false;
        }

        console.log('✅ Schema validation passed');

        // Validate business rules
        const businessErrors = validateBusinessRules(bank);
        if (businessErrors.length > 0) {
            console.error('❌ Business rule validation failed:');
            businessErrors.forEach(error => console.error(`  - ${error}`));
            return false;
        }

        console.log('✅ Business rule validation passed');

        // Summary
        const testletCount = bank.testlets?.length || 0;
        const questionCount = (bank.testlets || []).reduce((sum, t) => sum + (t.items?.length || 0), 0);
        const diagnosticCount = bank.diagnostics?.length || 0;

        console.log(`📊 Summary:`);
        console.log(`  - Testlets: ${testletCount}`);
        console.log(`  - Questions: ${questionCount}`);
        console.log(`  - Diagnostics: ${diagnosticCount}`);
        console.log('🎉 All validations passed!');

        return true;

    } catch (error) {
        console.error(`❌ Error: ${error.message}`);
        return false;
    }
}

// CLI execution
if (process.argv.length < 3) {
    console.error('Usage: node lint-bank.mjs <path-to-bank.json>');
    process.exit(1);
}

const bankPath = process.argv[2];
const success = await lintBank(bankPath);
process.exit(success ? 0 : 1);
